core/unit_test/TestHostSharedPtr.hpp

```diff
@@ -0,0 +1,247 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <impl/Kokkos_HostSharedPtr.hpp>
+
+#include <gtest/gtest.h>
+
+using Kokkos::Experimental::HostSharedPtr;
+using Kokkos::Experimental::UnmanagedPtr;
+
+TEST(TEST_CATEGORY, host_shared_ptr_is_reference_counted) {
+  using T = int;  // default constructible
+  {
+    HostSharedPtr<T> p1;
+    EXPECT_FALSE(p1.is_reference_counted());
+  }
+  {
+    HostSharedPtr<T> p1(nullptr);
+    EXPECT_FALSE(p1.is_reference_counted());
+  }
```

Comment:
I still think these should be `true`. Reference counting is an aspect of the type, not the value. It's a behavior, not an aspect of the data.

Reply:
If it were an aspect of the type, wouldn't it _always_ return `true`?

It would be weird if these were not all the same state:

- Default constructed
- Moved-from
- Constructed with a `nullptr`
- Constructed with a null pointer value

Reply:
> It would be weird if these were not all the same state:

Agreed

> And if 4 is required to have a deleter,

I don't see how that follows. `is_reference_counted()` doesn't imply there's a deleter, just that any object managed by the `MaybeSharedPtr` will have owning, reference-counted semantics.

> I prefer to follow what `unique_ptr` does here (fewer requirements on the `deleter`)

I agree.